internal/engine/compiler/impl_amd64.go

```diff
@@ -5013,3 +5082,43 @@ func (c *amd64Compiler) compileEnsureOnRegister(loc *runtimeValueLocation) (err
 	}
 	return
 }
+
+// compileMaybeSwapRegisters swaps two registers if they're not equal.
+func (c *amd64Compiler) compileMaybeSwapRegisters(reg1, reg2 asm.Register) {
+	if reg1 != reg2 {
+		c.assembler.CompileRegisterToRegister(amd64.XCHGQ, reg1, reg2)
+	}
+}
+
+// compilePreventCrossedTargetRegisters ensures that neither runtimeValueLocations target register
+// is present among the other runtimeValueLocations. To prevent this, it swaps misplaced registers.
+//
+// This function makes it possible to safely exchange one set of registers with another, where a register might be in both sets.
+// Each register will correspond either to another register not present in its set or to itself
+func (c *amd64Compiler) compilePreventCrossedTargetRegisters(locs []*runtimeValueLocation, targets []asm.Register) []int {
+	swaps := []int{}
+	for i := range locs {
+		targetLocation := -1 // -1 means not found.
+		for j := range locs {
+			if locs[j].register == targets[i] {
+				targetLocation = j
+				break
+			}
+		}
+		if targetLocation != -1 && targetLocation != i {
+			c.compileMaybeSwapRegisters(locs[i].register, locs[targetLocation].register)
+			locs[i].register, locs[targetLocation].register = locs[targetLocation].register, locs[i].register
+			swaps = append(swaps, i, targetLocation)
```

Comment:
I see  from this, `swaps []int` is more of `[]struct{srcIndex, dstIndex}`? Please use the named struct, rather than flatted []int as it's a bit complex to understand for others.